File: mobile/src/test/java/org/openhab/habdroid/core/connection/ConnectionFactoryTest.java
package org.openhab.habdroid.core.connection;

import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;
import org.mockito.Mockito;
import org.openhab.habdroid.core.connection.exception.ConnectionException;
import org.openhab.habdroid.core.connection.exception.NetworkNotAvailableException;
import org.openhab.habdroid.core.connection.exception.NetworkNotSupportedException;
import org.openhab.habdroid.core.connection.exception.NoUrlInformationException;
import org.openhab.habdroid.util.Constants;

import java.io.File;
import java.io.IOException;

import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.when;

public class ConnectionFactoryTest {
    @Rule
    public TemporaryFolder mTempFolder = new TemporaryFolder();

    private Context mMockContext;
    private ConnectivityManager mMockConnectivityService;
    private SharedPreferences mMockPrefs;

    @Before
    public void setup() throws IOException {
        mMockConnectivityService = Mockito.mock(ConnectivityManager.class);

        File cacheFolder = mTempFolder.newFolder("cache");
        File appDir = mTempFolder.newFolder();

        mMockContext = Mockito.mock(Application.class);
        Mockito.when(mMockContext.getApplicationContext()).thenReturn(mMockContext);
        Mockito.when(mMockContext.getCacheDir()).thenReturn(cacheFolder);
        Mockito.when(mMockContext.getDir(anyString(), anyInt()))
                .then(invocation -> new File(appDir, invocation.getArgument(0).toString()));
        when(mMockContext.getString(anyInt())).thenReturn("");
        when(mMockContext.getSystemService(eq(Context.CONNECTIVITY_SERVICE)))
                .thenReturn(mMockConnectivityService);
        when(mMockContext.getMainLooper()).thenReturn(Looper.getMainLooper());

        mMockPrefs = Mockito.mock(SharedPreferences.class);

        ConnectionFactory.initialize(mMockContext, mMockPrefs);

        ConnectionFactory.sInstance.mMainHandler = makeMockedHandler();
        ConnectionFactory.sInstance.mUpdateHandler = makeMockedHandler();
    }

    @Test
    public void testGetConnectionRemoteWithUrl() throws IOException {
        MockWebServer server = new MockWebServer();
        server.enqueue(new MockResponse().setResponseCode(404));
        server.start();

        when(mMockPrefs.getString(eq(Constants.PREFERENCE_REMOTE_URL), anyString()))
                .thenReturn(server.url("/").toString());
        ConnectionFactory.sInstance.updateConnections();
        Connection conn = ConnectionFactory.getConnection(Connection.TYPE_REMOTE);

        assertNotNull("Requesting a remote connection when a remote url is set, " +
                " should return a connection.", conn);
        assertEquals("The connection type of a remote connection should be TYPE_REMOTE.",
                Connection.TYPE_REMOTE, conn.getConnectionType());
    }

    @Test
    public void testGetConnectionRemoteWithoutUrl() {
        when(mMockPrefs.getString(eq(Constants.PREFERENCE_REMOTE_URL), anyString()))
                .thenReturn("");
        ConnectionFactory.sInstance.updateConnections();
        Connection conn = ConnectionFactory.getConnection(Connection.TYPE_REMOTE);

        assertNull("Requesting a remote connection when a remote url isn't set, should not " +
                "return a connection.", conn);
    }

    @Test
    public void testGetConnectionLocalWithUrl() {
        when(mMockPrefs.getString(eq(Constants.PREFERENCE_LOCAL_URL), anyString()))
                .thenReturn("https://openhab.local:8080");
        ConnectionFactory.sInstance.updateConnections();
        Connection conn = ConnectionFactory.getConnection(Connection.TYPE_LOCAL);

        assertNotNull("Requesting a local connection when local url is set, should " +
                "return a connection.", conn);
        assertEquals("The connection type of a local connection should be LOGLEVEL_LOCAL.",
                Connection.TYPE_LOCAL, conn.getConnectionType());
    }

    @Test
    public void testGetConnectionLocalWithoutUrl() {
        when(mMockPrefs.getString(eq(Constants.PREFERENCE_LOCAL_URL), anyString()))
                .thenReturn("");
        ConnectionFactory.sInstance.updateConnections();
        Connection conn = ConnectionFactory.getConnection(Connection.TYPE_LOCAL);

        assertNull("Requesting a remote connection when a local url isn't set, should not " +
                "return a connection.", conn);
    }

    @Test
    public void testGetConnectionCloudWithUrl() throws IOException {
        MockWebServer server = new MockWebServer();
        server.enqueue(new MockResponse().setBody("{'gcm': { 'senderId': '12345'} }"));
        server.start();

        when(mMockPrefs.getString(eq(Constants.PREFERENCE_REMOTE_URL), anyString()))
                .thenReturn(server.url("/").toString());

        ConnectionFactory.sInstance.updateConnections();
        Connection conn = ConnectionFactory.getConnection(Connection.TYPE_CLOUD);

        assertNotNull("Requesting a cloud connection when a remote url is set, "
                + "should return a connection.", conn);
        assertEquals(CloudConnection.class, conn.getClass());
        assertEquals("The connection type of a cloud connection should be TYPE_CLOUD.",
                Connection.TYPE_CLOUD, conn.getConnectionType());
        assertEquals("The sender ID of the cloud connection should be '12345'",
                "12345", ((CloudConnection) conn).getMessagingSenderId());

        server.shutdown();
    }

    @Test(expected = NetworkNotAvailableException.class)
    public void testGetAnyConnectionNoNetwork() throws ConnectionException {
        triggerNetworkUpdate(null);

        ConnectionFactory.getUsableConnection();
    }

    @Test(expected = NetworkNotSupportedException.class)
    public void testGetAnyConnectionUnsupportedNetwork() throws ConnectionException {
        triggerNetworkUpdate(ConnectivityManager.TYPE_BLUETOOTH);

        ConnectionFactory.getUsableConnection();
    }

    @Test
    public void testGetAnyConnectionWifiRemoteOnly() throws ConnectionException, IOException {
        MockWebServer server = new MockWebServer();
        server.enqueue(new MockResponse().setResponseCode(404));
        server.start();

        when(mMockPrefs.getString(eq(Constants.PREFERENCE_REMOTE_URL), anyString()))
                .thenReturn(server.url("/").toString());
        ConnectionFactory.sInstance.updateConnections();
        triggerNetworkUpdate(ConnectivityManager.TYPE_WIFI);

        Connection conn = ConnectionFactory.getUsableConnection();

        assertNotNull("Requesting any connection in WIFI when only a remote url is set, "
                + "should return a connection.", conn);
        assertEquals("The connection type of the connection should be TYPE_REMOTE.",
                Connection.TYPE_REMOTE, conn.getConnectionType());

        server.shutdown();
    }

    @Test
    public void testGetAnyConnectionWifiLocalRemote() throws ConnectionException, IOException {
        MockWebServer server = new MockWebServer();
        server.enqueue(new MockResponse().setResponseCode(404));
        server.start();

        when(mMockPrefs.getString(eq(Constants.PREFERENCE_REMOTE_URL), anyString()))
                .thenReturn(server.url("/").toString());
        when(mMockPrefs.getString(eq(Constants.PREFERENCE_LOCAL_URL), anyString()))
                .thenReturn("https://myopenhab.org:443");
        ConnectionFactory.sInstance.updateConnections();
        triggerNetworkUpdate(ConnectivityManager.TYPE_WIFI);

        Connection conn = ConnectionFactory.getUsableConnection();

        assertNotNull("Requesting any connection in WIFI when a local url is set, "
                + "should return a connection.", conn);
        assertEquals("The connection type of the connection should be TYPE_LOCAL.",
                Connection.TYPE_LOCAL, conn.getConnectionType());

        server.shutdown();
    }

    @Test(expected = NoUrlInformationException.class)
    public void testGetAnyConnectionWifiNoLocalNoRemote() throws ConnectionException {
        when(mMockPrefs.getString(anyString(), anyString())).thenReturn(null);
        triggerNetworkUpdate(ConnectivityManager.TYPE_WIFI);

        ConnectionFactory.getUsableConnection();
    }

    private void triggerNetworkUpdate(int type) {
        NetworkInfo mockNetworkInfo = Mockito.mock(NetworkInfo.class);
        when(mockNetworkInfo.getType()).thenReturn(type);
        when(mockNetworkInfo.isConnected()).thenReturn(true);
        triggerNetworkUpdate(mockNetworkInfo);
    }

    private void triggerNetworkUpdate(NetworkInfo info) {
        when(mMockConnectivityService.getActiveNetworkInfo()).thenReturn(info);

        ConnectionFactory.sInstance.onReceive(mMockContext,
                new Intent(ConnectivityManager.CONNECTIVITY_ACTION));
    }

    private Handler makeMockedHandler() {
        final Handler h = Mockito.mock(Handler.class);
        when(h.sendEmptyMessage(anyInt())).thenAnswer(invocation -> {
            Message msg = new Message();
            msg.what = invocation.getArgument(0);
            ConnectionFactory.sInstance.handleMessage(msg);
            return Boolean.TRUE;
        });
        when(h.sendMessage(any(Message.class))).thenAnswer(invocation -> {
            Message msg = invocation.getArgument(0);
            ConnectionFactory.sInstance.handleMessage(msg);
            return Boolean.TRUE;
        });
        when(h.obtainMessage(anyInt()))
                .thenAnswer(invocation -> makeMockedMessage(h, invocation.getArgument(0), null));
        when(h.obtainMessage(anyInt(), any()))
                .thenAnswer(invocation -> makeMockedMessage(h, invocation.getArgument(0), invocation.getArgument(1)));
        return h;
    }

    private Message makeMockedMessage(Handler h, int what, Object obj) {
        Message msg = Mockito.mock(Message.class);
        msg.what = what;
        msg.obj = obj;
        doAnswer(invocationOnMock -> h.sendMessage(msg)).when(msg).sendToTarget();
        return msg;
    }
}

File: mobile/src/test/java/org/openhab/habdroid/core/connection/ConnectionFactoryTest.kt
/*
 * Copyright (c) 2010-2024 Contributors to the openHAB project
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package org.openhab.habdroid.core.connection

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.net.Network
import android.net.NetworkCapabilities
import android.net.wifi.WifiManager
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.anyOrNull
import com.nhaarman.mockitokotlin2.doAnswer
import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import java.io.File
import java.io.IOException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import kotlinx.coroutines.newSingleThreadContext
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.setMain
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.AfterClass
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.BeforeClass
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import org.openhab.habdroid.testUtils.Retry
import org.openhab.habdroid.testUtils.RetryRule
import org.openhab.habdroid.util.PrefKeys

class ConnectionFactoryTest {
    @ExperimentalCoroutinesApi
    companion object {
        private val mainThread = newSingleThreadContext("UI thread")

        @AfterClass
        @JvmStatic
        fun tearDownMainThread() {
            Dispatchers.resetMain()
            mainThread.close()
        }

        @BeforeClass
        @JvmStatic
        @Throws(IOException::class)
        fun setupMainThread() {
            Dispatchers.setMain(mainThread)
        }
    }

    @Rule
    @JvmField
    val tempFolder = TemporaryFolder()

    @Rule
    @JvmField
    val retry = RetryRule()

    private lateinit var mockWifiManager: WifiManager
    private lateinit var mockContext: Application
    private lateinit var mockPrefs: SharedPreferences
    private lateinit var mockNetwork: Network
    private lateinit var mockNetworkCaps: NetworkCapabilities
    private val mockConnectionHelper = MockConnectionHelper()

    @Before
    @Throws(IOException::class)
    fun setup() {
        val cacheFolder = tempFolder.newFolder()
        val appDir = tempFolder.newFolder()

        mockPrefs = mock {
            on { getStringSet(eq(PrefKeys.SERVER_IDS), anyOrNull()) } doReturn setOf("1")
            on { getInt(eq(PrefKeys.ACTIVE_SERVER_ID), any()) } doReturn 1
        }

        mockWifiManager = mock {
        }
        @Suppress("DEPRECATION")
        whenever(mockWifiManager.connectionInfo) doReturn null

        mockContext = mock<Application> {
            on { cacheDir } doReturn cacheFolder
            on { getDir(any(), any()) } doAnswer { invocation ->
                File(appDir, invocation.getArgument<Any>(0).toString())
            }
            on { getString(any()) } doReturn ""
            on { getSystemService(Context.WIFI_SERVICE) } doReturn mockWifiManager
        }
        whenever(mockContext.applicationContext) doReturn mockContext

        mockNetwork = mock {}
        mockNetworkCaps = mock {}

        ConnectionFactory.initialize(mockContext, mockPrefs, mockConnectionHelper)
    }

    @Test
    @Retry
    @Throws(IOException::class)
    fun testGetConnectionRemoteWithUrl() {
        val server = MockWebServer()
        server.enqueue(MockResponse().setResponseCode(404))
        server.start()

        fillInServers(remote = server.url("/").toString())
        updateAndWaitForConnections()
        assertTrue(
            "Should return a connection if remote url is set.",
            ConnectionFactory.hasActiveRemoteConnection
        )
    }

    @Test
    @Retry
    fun testGetConnectionRemoteWithoutUrl() {
        fillInServers(remote = "")
        updateAndWaitForConnections()
        assertFalse(
            "Should not return a remote connection if remote url isn't set.",
            ConnectionFactory.hasActiveRemoteConnection
        )
    }

    @Test
    @Retry
    fun testGetConnectionLocalWithUrl() {
        fillInServers(local = "https://openhab.local:8080")
        updateAndWaitForConnections()
        assertTrue(
            "Should return a local connection if local url is set.",
            ConnectionFactory.hasActiveLocalConnection
        )
    }

    @Test
    @Retry
    fun testGetConnectionLocalWithoutUrl() {
        fillInServers(local = "")
        updateAndWaitForConnections()
        assertFalse(
            "Should not return a local connection when local url isn't set.",
            ConnectionFactory.hasActiveLocalConnection
        )
    }

    @Test
    @Retry
    @Throws(IOException::class)
    fun testGetConnectionCloudWithUrl() {
        val server = MockWebServer()
        server.enqueue(MockResponse().setBody("{'gcm': { 'senderId': '12345'} }"))
        server.start()

        fillInServers(remote = server.url("/").toString())
        updateAndWaitForConnections()
        val conn = ConnectionFactory.activeCloudConnection?.connection

        assertNotNull("Should return a cloud connection if remote url is set.", conn)
        assertEquals(CloudConnection::class.java, conn!!.javaClass)
        assertEquals(
            "The connection type of a cloud connection should be TYPE_CLOUD.",
            Connection.TYPE_CLOUD,
            conn.connectionType
        )
        assertEquals(
            "The sender ID of the cloud connection should be '12345'",
            "12345",
            conn.messagingSenderId
        )

        server.shutdown()
    }

    @Test
    @Retry
    fun testGetAnyConnectionNoNetwork() {
        mockConnectionHelper.update(null)
        updateAndWaitForConnections()
        assertEquals(
            ConnectionFactory.activeUsableConnection?.failureReason?.javaClass,
            NetworkNotAvailableException::class.java
        )
    }

    @Test
    @Retry
    fun testGetConnectionUnknownNetwork() {
        fillInServers("https://openhab.local:8080", "https://openhab.local:8080")
        mockConnectionHelper.update(ConnectionManagerHelper.ConnectionType.Unknown(mockNetwork, mockNetworkCaps))
        updateAndWaitForConnections()
        assertEquals(
            "Unknown transport types should be used for remote connections",
            ConnectionFactory.activeUsableConnection?.connection?.connectionType,
            Connection.TYPE_REMOTE
        )
    }

    @Test
    @Retry
    fun testGetAnyConnectionWifiRemoteOnly() {
        val server = MockWebServer()
        server.enqueue(MockResponse().setResponseCode(404))
        server.start()

        fillInServers(remote = server.url("/").toString())
        mockConnectionHelper.update(ConnectionManagerHelper.ConnectionType.Wifi(mockNetwork, mockNetworkCaps))
        updateAndWaitForConnections()

        val conn = ConnectionFactory.activeUsableConnection?.connection

        assertNotNull("Should return a connection in WIFI when only remote url is set.", conn)
        assertEquals(
            "The connection type of the connection should be TYPE_REMOTE.",
            Connection.TYPE_REMOTE,
            conn?.connectionType
        )

        server.shutdown()
    }

    @Test
    @Retry
    fun testGetAnyConnectionWifiLocalRemote() {
        val server = MockWebServer()
        server.enqueue(MockResponse().setResponseCode(404))
        server.start()

        fillInServers(remote = server.url("/").toString(), local = "https://myopenhab.org:443")
        mockConnectionHelper.update(ConnectionManagerHelper.ConnectionType.Wifi(mockNetwork, mockNetworkCaps))
        updateAndWaitForConnections()

        val conn = ConnectionFactory.activeUsableConnection?.connection

        assertNotNull("Should return a connection in WIFI when a local url is set.", conn)
        assertEquals(
            "The connection type of the connection should be TYPE_LOCAL.",
            Connection.TYPE_LOCAL,
            conn?.connectionType
        )

        server.shutdown()
    }

    @Test
    @Retry
    fun testGetAnyConnectionWifiNoLocalNoRemote() {
        fillInServers(null, null)
        mockConnectionHelper.update(ConnectionManagerHelper.ConnectionType.Wifi(mockNetwork, mockNetworkCaps))
        updateAndWaitForConnections()
        assertEquals(
            ConnectionFactory.activeUsableConnection?.failureReason?.javaClass,
            NoUrlInformationException::class.java
        )
    }

    private inner class MockConnectionHelper : ConnectionManagerHelper {
        override var changeCallback: ConnectionChangedCallback? = null
        private var currentTypes: List<ConnectionManagerHelper.ConnectionType> = emptyList()
        override val currentConnections: List<ConnectionManagerHelper.ConnectionType> get() = currentTypes

        fun update(type: ConnectionManagerHelper.ConnectionType?) {
            currentTypes = if (type != null) listOf(type) else emptyList()
            changeCallback?.invoke()
        }

        override fun start() {}
        override fun shutdown() {}
    }

    private fun fillInServers(local: String? = null, remote: String? = null) {
        whenever(
            mockPrefs.getString(eq(PrefKeys.buildServerKey(1, PrefKeys.LOCAL_URL_PREFIX)), anyOrNull())
        ) doReturn local
        whenever(
            mockPrefs.getString(eq(PrefKeys.buildServerKey(1, PrefKeys.REMOTE_URL_PREFIX)), anyOrNull())
        ) doReturn remote
        whenever(
            mockPrefs.getString(eq(PrefKeys.buildServerKey(1, PrefKeys.SERVER_NAME_PREFIX)), anyOrNull())
        ) doReturn "Test Server"
    }

    private fun updateAndWaitForConnections() {
        runBlocking {
            launch(Dispatchers.Main) {
                ConnectionFactory.instance.updateConnections()
            }
            ConnectionFactory.waitForInitialization()
        }
    }
}
